In [1]:
# Functions
#Q1 Create a function to calculate the the area of triangle

def Area(base, height):
    return 0.5 * base * height
print(Area(10, 12))

60.0


In [2]:
#Q2 Create a function that calculates the the volume of sphere

def vol(r):
    return (4/3)*3.14*(r**3)
print(vol(5))

523.3333333333334


In [4]:
#Q3 Explain what is default and required arguments with an example of each

# Default values are passed using '=' operator.

def areaofcircle(radius,pi=3.14):
    return pi*radius*radius
print (areaofcircle(4))
# pi is default argument

# required arguments must be passed during the fuction call 
# and exactly in the right order.

def areaofrectangle(length,width):
    return length * width
print(areaofrectangle(20,25))


50.24
500


In [9]:
#Q4 Explain what is variable length arguments i.e *args & **kwargs with an example of each

#Arbitrary Arguments, *args

#If you do not know how many arguments that will be passed into your function, add a * before the parameter
#name in the function definition.

def child(*kids):
    print('The youngest child is '+ kids[2])
child('a','b','c')

#Arbitrary Keyword Arguments, **kwargs

#If you do not know how many keyword arguments that will be passed into your function, add two asterisk: **
#before the parameter name in the function definition.

def child(**girl):
    print('Her name is '+ girl['fn'])
child(fn='chandni',ln='arya')

The youngest child is c
Her name is chandni


In [13]:
#Q5 WAP to create function of prime no
import math

def checkPrime(n):
    if n < 2:
        return 1
    for i in range(2, int(math.sqrt(n))+1):
        if n%i == 0:
            return 1
    return 0

print("Enter a Number: ", end="")
n = int(input())

p = checkPrime(n)
if p == 1:
    print(n, "is not a Prime Number")
else:
    print(n, "is a Prime Number")

    
#The given code defines a function "checkPrime(n)"" that takes an integer n as input and checks if it is a prime number or not.

#The function returns "1" if the number is not prime and 0 if it is prime.

#The function first checks if the number is less than "2".

#If it is, then it is not a prime number, so the function returns "1".

#If the number is greater than or equal to "2", the function checks if it is divisible by any number from "2" to the square root of "n". 

#If it is divisible by any number in that range, then it is not a prime number, so the function returns "1". 

#Otherwise, it returns "0", indicating that the number is prime.

#After defining the checkPrime(n) function, the code prompts the user to enter a number and assigns it to the variable "n".

#The code then calls the checkPrime(n) function to check if the number is prime or not.

#If the function returns "1", the code prints a message indicating that the number is not prime.

#Otherwise, it prints a message indicating that the number is prime.

Enter a Number: 5
5 is a Prime Number


In [10]:
#Q6 WAP to create a function of perfect number

def perfect_number(n):
    sum = 0
    for i in range(1, n):
        if n % i == 0:
            sum = sum+i 
    return sum 
n = int(input('Enter Number: '))
if(n == perfect_number(n)): 
    print(n, "is a perfect number") 
else: 
    print(n, "is not a perfect number") 
    
# The code defines a function perfect_number(n) that takes an integer n as input and computes the sum of all the proper divisors of n.
# The function returns the sum.
# The function uses a loop that iterates over all the numbers from 1 to n-1. For each number i, it checks if n is divisible by i.
# If it is, then i is a proper divisor of n, so it adds i to the sum.
# After defining the perfect_number(n) function, the code prompts the user to enter a The code then calls the perfect_number(n) function to compute the sum of the proper divisors of n.
# If the sum is equal to n, the code prints a message indicating that n is a perfect number. 
# Otherwise, it prints a message indicating that n is not a perfect number.
# A perfect number is a positive integer that is equal to the sum of its proper divisors.
# For example, 6 is a perfect number because its proper divisors are 1, 2, and 3, and 1 + 2 + 3 = 6.
# Note that the code assumes that the input n is a positive integer. 
# If the user enters a non-positive integer, the behavior of the code is not well-defined.

Enter Number: 6
6 is a perfect number


In [2]:
#Q7 WAP to create a function of neon number

def neon(x) :
    sqr = x*x
    sum = 0
    while(sqr!=0):
        sum=sum+(sqr%10)
        sqr=sqr//10
    return sum==x

i = 1
while i <= 10 : 
    if (neon(i)) : 
        print (i,end=' ')
    i = i + 1   
    
#This is a Python code that defines a function neon(x) which takes an integer x as input and returns True if the sum of the digits of the square of x is equal to x, otherwise it returns False.

#Then, a loop is used to iterate over the integers from 1 to 10, and for each integer i, the neon() function is called with i as input. 
#If the function returns True, then i is printed.

#So, when you run this code, the output will be:
#1,9
#1 is a neon number because its square is 1, and the sum of the digits of 1 is also 1.
#9 is a neon number because its square is 81, and the sum of the digits of 81 is 9.
#Note that the output only includes the neon numbers between 1 and 10 (inclusive). 
#If you want to find all the neon numbers between 1 and a larger number n,
# you can modify the loop to iterate over all the integers from 1 to n, and print the ones for which neon() returns True.







1 9 

In [7]:
#Q8 Explain what is lambda function with the help of example
#anonymous functions are also called lambda functions 
#lambda arguments: expression

list1=[2,5,8,6,10,12,13,11]
odd_number=list(filter(lambda x: x % 2!=0,list1))
print("Odd numbers are: ", odd_number)

#This is a Python code that defines a list of integers list1, and then applies the filter() function to this list using a lambda function that checks if the number is odd or not.

#The filter() function takes two arguments: a function that returns a boolean value, and an iterable (in this case, the list list1). 
#It returns a new iterable that contains only the elements of the original iterable for which the function returns True.

#So, in this code, the lambda function lambda x: x % 2 != 0 checks if x is odd (i.e., if x % 2 is not equal to 0).

#The filter() function applies this lambda function to each element of list1 and returns a new iterable containing only the odd numbers.
#This new iterable is stored in the variable odd_number.

#Finally, the code prints a message that displays the odd numbers by using the print() function to concatenate the string "Odd numbers are: " with the list of odd numbers (odd_number) that was obtained from the filter() function.

#When you run this code, the output will be:

Odd numbers are:  [5, 13, 11]


In [2]:
#Q9 WAP for factorial using recurrsion

def factorial(no):
    if no == 0:
        return 1
    else:
        return no * factorial(no - 1)
print("factorial of a number is:", factorial(9))



factorial of a number is: 362880


This is a Python code that defines a recursive function factorial(no) that takes an integer no as input and returns its factorial.
In mathematics, the factorial of a non-negative integer n is the product of all positive integers less than or equal to n. 
By convention, the factorial of 0 is 1.
The function factorial(no) uses recursion to compute the factorial of no.
If no is 0, the function returns 1 (since the factorial of 0 is 1). 
Otherwise, it computes the factorial of no-1 by calling itself recursively with no-1 as the argument, and multiplies the result by no.
#In this code, the function factorial(9) is called to compute the factorial of 9. 
When you run this code, the output will be: 362880

explanation:
The function factorial(9) is called to compute the factorial of 9.
Since 9 is not 0, the function computes the factorial of 8 by calling itself recursively with 8 as the argument, and multiplies the result by 9.
Since 8 is not 0, the function computes the factorial of 7 by calling itself recursively with 7 as the argument,
and multiplies the result by 8. This process continues until no becomes 0, at which point the function returns 1.
Finally, the result (362880) is printed using the print() function.

In [19]:
#Q10 WAP for fibonacci series using recursion using one of the memoization technique

nums={}
def fibonacci(n):
    if n<=2:
        return 1
    if n in nums:
        return nums[n]
    else:
        return fibonacci(n-1)+fibonacci(n-2)
        nums[n]=num
        return num
    
print(fibonacci(10))



55


The function fibonacci(n) uses recursion to compute the n-th Fibonacci number. 
If n is less than or equal to 2, the function returns 1.
Otherwise, it checks if the value of n has already been computed and stored in a dictionary called nums.
This is a Python code that defines a recursive function fibonacci(n) that takes an integer n as input and returns the n-th Fibonacci number.
In the Fibonacci sequence, each number is the sum of the two preceding ones, starting from 0 and 1.
That is, the first two numbers in the Fibonacci sequence are 0 and 1, an
each subsequent number is the sum of the previous two.
The first few numbers in the sequence are: 0, 1, 1, 2, 3, 5, 8, 13, 21, ....
If it has, the function returns the value from the dictionary.
If it has not, the function computes the value recursively as the sum of the n-1-th and n-2-th Fibonacci numbers,
stores the result in the dictionary, and returns it
In this code, the fibonacci(10) function is called to compute the 10th Fibonacci number. 

#EXPLANATION:
The function fibonacci(10) is called to compute the 10th Fibonacci number.
The function computes the value of the 10th Fibonacci number recursively as the sum of the 9th and 8th Fibonacci numbers.
Since the 9th and 8th Fibonacci numbers have not yet been computed, the function recursively computes them as 
the sum of the 8th and 7th, and the 7th and 6th Fibonacci numbers, respectively.
This process continues until the function computes the 1st and 2nd Fibonacci numbers, which are both 1.
The function stores the computed Fibonacci numbers in a dictionary called nums to avoid recomputing them in the future.
Finally, the function returns the value of the 10th Fibonacci number, which is 55.


In [5]:
#Q11.Item with highest value count

items = [{'apple': 5}, {'banana': 8}, {'orange': 7}, {'grapes': 4}]
def itemWithHighestValue(items):
    pass
z={}
for d in items:
    z.update(d)
z.items()
s=sorted(z.items(),key=lambda tup:tup[1])
s[-1]



('banana', 8)

The code above is a Python script that finds the item with the highest value in a list of dictionaries items. Here's what the code does step by step:

items = [{'apple': 5}, {'banana': 8}, {'orange': 7}, {'grapes': 4}]: This line creates a list of dictionaries called items.

def itemWithHighestValue(items): pass: This line defines a function called itemWithHighestValue that takes a list of dictionaries items as input but does nothing.

z = {}: This line initializes an empty dictionary called z.

for d in items: z.update(d): This line loops through each dictionary d in the items list and updates the z dictionary with its key-value pairs.

z.items(): This line returns a list of the key-value pairs in the z dictionary.

s = sorted(z.items(), key=lambda tup: tup[1]): This line sorts the key-value pairs in the z dictionary by their values in ascending order and stores the sorted list in s.

s[-1]: This line returns the last (highest) item in the s list, which is the item with the highest value.

So, the output of the script will be the key-value pair of the item with the highest value. For example, if the items list is:
[{'apple': 5}, {'banana': 8}, {'orange': 7}, {'grapes': 4}]
Then the output of the script will be:

('banana', 8)
**Note that the code assumes that the value of each item is a numeric value that can be sorted in ascending order. If the values are not numeric or cannot be sorted, the code may not work as expected. Also note that the function itemWithHighestValue is not used in the code and can be removed.

In [5]:
#Q12 Create a lambda function to sort the name list according to their surname.

name = ["jay A kumar",'deepak N sharma','rohit N yadav','mithilesh singh','payal vasave','Fahad Momin']

name.sort(key=lambda x: x.split(" ")[-1].lower())
print(name)



['jay A kumar', 'Fahad Momin', 'deepak N sharma', 'mithilesh singh', 'payal vasave', 'rohit N yadav']


#name is a list of strings that contains the names to be sorted.

#sort() is a built-in Python method used to sort the elements of a list in ascending order.

 The "key" parameter is set to a lambda function that extracts the last name of each name and converts it to lowercase before sorting.

split(" ") is used to split each name into a list of words using space as the delimiter.

 [-1] is used to select the last element of the list, which is the last name.

 lower() is used to convert the last name to lowercase before sorting.

#So, the output of the code would be a sorted list of names based on their last names in ascending order.


In [22]:
#Q13 using map function, create a lambda expression to convert the degree celcius to fahrenheit
 #f= 9/5 x c + 32

temp = [('kashmir',1),('mumbai',16),('goa',22),('channai',18),('kerla',20)]
places = list(map(lambda c: (c[0], (9/5) * c[1]+ 32), temp))
print(places)



[('kashmir', 33.8), ('mumbai', 60.8), ('goa', 71.6), ('channai', 64.4), ('kerla', 68.0)]


The code above creates two lists, name and roll_no, and then uses the zip() function to create a list of tuples where each tuple contains a name and its corresponding roll number. Here's what the code does step by step:

name = [ "Manjeet", "Nikhil", "Shambhavi", "Astha" ]: This line creates a list of strings containing some names.

roll_no = [ 4, 1, 3, 2 ]: This line creates a list of integers containing some roll numbers.

mapped = zip(name, roll_no): This line uses the zip() function to combine the name and roll_no lists into a single list of tuples where each tuple contains a name and its corresponding roll number.

mapped = list(mapped): This line converts the mapped object (which is a zip object) into a list so that we can print its contents.

print(mapped): This line prints out the list of tuples containing the names and roll numbers.

So, the output of the script will be:

css
Copy code
[('Manjeet', 4), ('Nikhil', 1), ('Shambhavi', 3), ('Astha', 2)]
Note that the order of the tuples in the resulting list depends on the order of the elements in the original lists. In this case, the names and roll numbers are in the order they were defined in the original lists.

In [6]:
#Q14 Give example of zip(), abs(), round(),pow() function

# a) zip()
name = [ "Manjeet", "Nikhil", "Shambhavi", "Astha" ]
roll_no = [ 4, 1, 3, 2 ]
mapped = zip(name, roll_no)
mapped = list(mapped)
print(mapped)

[('Manjeet', 4), ('Nikhil', 1), ('Shambhavi', 3), ('Astha', 2)]


The code above creates two lists, name and roll_no, and then uses the zip() function to create a list of tuples where each tuple contains a name and its corresponding roll number. Here's what the code does step by step:

name = [ "Manjeet", "Nikhil", "Shambhavi", "Astha" ]: This line creates a list of strings containing some names.

roll_no = [ 4, 1, 3, 2 ]: This line creates a list of integers containing some roll numbers.

mapped = zip(name, roll_no): This line uses the zip() function to combine the name and roll_no lists into a single list of tuples where each tuple contains a name and its corresponding roll number.

mapped = list(mapped): This line converts the mapped object (which is a zip object) into a list so that we can print its contents.

print(mapped): This line prints out the list of tuples containing the names and roll numbers.

So, the output of the script will be:

css
Copy code
[('Manjeet', 4), ('Nikhil', 1), ('Shambhavi', 3), ('Astha', 2)]
Note that the order of the tuples in the resulting list depends on the order of the elements in the original lists. In this case, the names and roll numbers are in the order they were defined in the original lists.

In [8]:
#b) abs() (absolute) distance from 0 to 2nd position

abs(-2)


2

In [9]:
#c) round()

round(4.8765,2)

4.88

In [14]:
#d) pow() power 5**5

pow(5,5)

3125

In [2]:
#Q15 a) apply filter to create a list of palindrome numbers from mylist

mylist = [11,14,5,78,45,67,9,88,10,303]

n=list(filter(lambda x:str(x)==str(x)[::-1],mylist))

print(n)

[11, 5, 9, 88, 303]


The code above is a Python script that filters out the palindrome numbers from a given list mylist. Here's what the code does step by step:

mylist = [11,14,5,78,45,67,9,88,10,303]: This line creates a list of integers called mylist.

lambda x:str(x)==str(x)[::-1]: This is a lambda function that takes an integer x as input, converts it to a string using str(x), and checks if the string is equal to its reverse (i.e., if the number is a palindrome).

filter(lambda x:str(x)==str(x)[::-1],mylist): This line applies the lambda function to each element in the mylist list using the filter() function. The filter() function returns an iterator that only contains the elements of the input sequence for which the lambda function returns True.

n=list(...): This line converts the result of the filter() function to a list and assigns it to the variable n.

print(n): This line prints the contents of the n list.

So, the output of the script will be a list of integers that are palindrome numbers from the mylist list. For example:


In [4]:
#Q15 b)apply filter to create a list of numbers divisible by 3 or 5 from the mylist


mylist = [11,14,5,78,45,67,9,88,10,303]
result = list(filter(lambda x: (x % 3 == 0 or x % 5 == 0 ), mylist))
print("Numbers divisible by 3 or 5 are",result)

Numbers divisible by 3 or 5 are [5, 78, 45, 9, 10, 303]


The code above is a Python script that filters out the numbers from a given list mylist that are divisible by 3 or 5. Here's what the code does step by step:

mylist = [11,14,5,78,45,67,9,88,10,303]: This line creates a list of integers called mylist.

lambda x: (x % 3 == 0 or x % 5 == 0 ): This is a lambda function that takes an integer x as input, and checks if the integer is divisible by either 3 or 5. If it is divisible, the lambda function returns True, otherwise it returns False.

filter(lambda x: (x % 3 == 0 or x % 5 == 0 ), mylist): This line applies the lambda function to each element in the mylist list using the filter() function. The filter() function returns an iterator that only contains the elements of the input sequence for which the lambda function returns True.

result = list(...): This line converts the result of the filter() function to a list and assigns it to the variable result.

print("Numbers divisible by 3 or 5 are",result): This line prints a message along with the contents of the result list.

So, the output of the script will be a list of integers that are divisible by either 3 or 5 from the mylist list. For example:

csharp
Copy code
Numbers divisible by 3 or 5 are [5, 78, 45, 9, 10, 303]
Note that the numbers 5, 78, 45, 9, 10, and 303 are divisible by either 3 or 5. The other numbers in the mylist list are not divisible by either 3 or 5.

In [7]:
#Q16 WAP to find maximum number from the list without using function and slicing or sorting
    #function
    
mylist = [11,14,5,78,45,67,9,88,10,303]
newlist = []

while mylist:
    minimum = mylist[0]  
    for i in mylist: 
        if i < minimum:
            minimum = i
    newlist.append(minimum)
    mylist.remove(minimum) 
print (newlist)

[5, 9, 10, 11, 14, 45, 67, 78, 88, 303]


The code above is a Python script that sorts a given list mylist in ascending order using the selection sort algorithm. Here's what the code does step by step:

mylist = [11,14,5,78,45,67,9,88,10,303]: This line creates a list of integers called mylist.

newlist = []: This line creates an empty list called newlist that will be used to store the sorted list.

while mylist:: This line starts a while loop that will execute as long as the mylist list is not empty.

minimum = mylist[0]: This line initializes the variable minimum to the first element of the mylist list.

for i in mylist:: This line starts a for loop that iterates over each element in the mylist list.

if i < minimum:: This line checks if the current element i is less than the current minimum value stored in the minimum variable.

minimum = i: This line updates the minimum variable to the new minimum value if i is less than the current minimum value.

newlist.append(minimum): This line appends the current minimum value to the newlist list.

mylist.remove(minimum): This line removes the current minimum value from the mylist list.

print(newlist): This line prints the sorted list stored in the newlist variable.

So, the output of the script will be a sorted list of integers in ascending order.
For example:
[5, 9, 10, 11, 14, 45, 67, 78, 88, 303]
**Note that the code uses the selection sort algorithm to sort the list. The selection sort algorithm works by finding the smallest element in the list and moving it to the front, then repeating the process with the remaining elements until the list is sorted.

In [10]:
def name(*name):
    print("Hello", name[0], name[1], name[2])
name("James", "Buchanan", "Barnes")
name("Abhishek","Kunal","Vinita")

Hello James Buchanan Barnes
Hello Abhishek Kunal Vinita


The code above is a Python script that defines a function called name that takes in a variable number of arguments using the *args syntax. Here's what the code does step by step:

def name(*name):: This line defines a function called name that takes in a variable number of arguments and stores them in a tuple called name.

print("Hello", name[0], name[1], name[2]): This line prints out a greeting message that includes the first three arguments passed to the function.

name("James", "Buchanan", "Barnes"): This line calls the name function with three arguments: "James", "Buchanan", and "Barnes". The function then prints out a greeting message that includes "James", "Buchanan", and "Barnes".

name("Abhishek", "Kunal", "Vinita"): This line calls the name function with three different arguments: "Abhishek", "Kunal", and "Vinita". The function then prints out a greeting message that includes "Abhishek", "Kunal", and "Vinita".

So, the output of the script will be:

Hello James Buchanan Barnes
Hello Abhishek Kunal Vinita
**Note that the name function assumes that there will always be at least three arguments passed to it. If the number of arguments is less than three or more than three, the function may produce an error.

In [4]:
def name(**name)
    print("Hello,", name["fname"], name["mname"], name["lname"])

name(mname = "Buchanan", lname = "Barnes", fname = "James")

Hello, James Buchanan Barnes


the code above defines a function called name that takes in a variable number of keyword arguments using the **kwargs syntax. Here's what the code does step by step:

def name(**name):: This line defines a function called name that takes in a variable number of keyword arguments and stores them in a dictionary called name.

print("Hello,", name["fname"], name["mname"], name["lname"]): This line prints out a greeting message that includes the values of the "fname", "mname", and "lname" keys in the name dictionary.

name(mname = "Buchanan", lname = "Barnes", fname = "James"): This line calls the name function with three keyword arguments: "fname" with a value of "James", "mname" with a value of "Buchanan", and "lname" with a value of "Barnes". The function then prints out a greeting message that includes "James", "Buchanan", and "Barnes".

So, the output of the script will be:

Hello, James Buchanan Barnes
Note that the order in which the keyword arguments are passed to the function does not matter, since they are stored in a dictionary and accessed using their respective keys.

In [11]:
num = input("Enter a number: ")
if num == num[::-1]:
    print("The number is a palindrome.")
else:
    print("The number is not a palindrome.")

Enter a number: 153
The number is not a palindrome.


The code above prompts the user to enter a number and then checks if the number is a palindrome. Here's what the code does step by step:

num = input("Enter a number: "): This line prompts the user to enter a number and stores it as a string in the num variable.

if num == num[::-1]:: This line checks if the num string is equal to its reverse. The [::-1] syntax is used to reverse the string.

print("The number is a palindrome."): This line is executed if the num string is a palindrome. It prints out a message saying that the number is a palindrome.

else:: This line is executed if the num string is not a palindrome.

print("The number is not a palindrome."): This line prints out a message saying that the number is not a palindrome.

So, if the user enters a palindrome number like "121", the output will be:

Enter a number: 121
The number is a palindrome.
If the user enters a non-palindrome number like "123", the output will be:

Enter a number: 123
The number is not a palindrome.
